### ISSUE TITLE
Metrics for unsupervised segmentation

### ISSUE THREAD

**Issue by Bralio123** [xu-ji/IIC#34]
Hi @xu-ji 

I managed to use IID for segmentation on Potsdam (3 and 6) but was wondering how you performed objective measures, e.g. pixel accuracy, to compare my results with yours. 

Do you map ground-truth masks to cluster predictions taken from a network maximising IID?

Thanks in advance.

**Comment by xu-ji** [comment]
Let C be the set of clusters, K be the set of ground truth classes for your labelled test set. |C| = |K|. In the case of segmentation, we do not operate at the level of image masks, but instead treat every 1*|C| output at a pixel as a single prediction. 

There is an associated 1*|K| ground truth vector for each pixel. We find the best 1-1 mapping between C and K using the hungarian algorithm (linear assignment), that function is [here](https://github.com/xu-ji/IIC/blob/master/code/utils/cluster/eval_metrics.py#L29). Then we re-order the output channels of the heatmap according to that mapping, and take accuracy (total number of correct pixels divided by all pixels) as normal.

It's actually just the same procedure as the whole image clustering evaluation, but each image gives h*w predictions instead of 1. There is also the detail that the mapping should be found with the training set for fairness, which we do follow because here the test set is (the labelled part of) the training set.

**Comment by Bralio123** [comment]
Thanks for the helpful response. I'll close this issue.

**Comment by Bralio123** [comment]
Hi again @xu-ji ,

Could you just help me with the inputs to the Hungarian match function you linked me [(function)](https://github.com/xu-ji/IIC/blob/ae173d08df6c6108ceace5dd5f6795183dab9f63/code/utils/cluster/eval_metrics.py).

Are the input tensors softmax preds of a given pixel, e.g. 1x3 for Potsdam 3, or are these 1x3 (again for Potsdam 3) OHE tensors with 1 encoding the max likelihood for a given pixel?

If its the latter, than using the linked function returns a list of tuple, which I imagine are the output channels mappings that need to be re-arranged. 

Thanks in advance,
Brandon

**Comment by xu-ji** [comment]
The inputs to `_hungarian_match` are the indices of each top cluster (argmax). E.g. for batch size 3 with 5 clusters,`flat_preds` could look like [3, 0, 4]. Not [[0, 0, 0, 1, 0], [1, 0, 0, 0, 0], [0, 0, 0, 0, 1]] or [[0.15, 0.15, 0.15, 0.4, 0.15], [0.4, 0.15, 0.15, 0.15, 0.15], [0.15, 0.15, 0.15, 0.15, 0.4]].

The return is a list of tuples of `(out_c, gt_c)` which is the one-to-one association between output cluster indices `out_c` and ground truth indices `gt_c`. You'd use this to swap the channel order of the output predictions so that the one mapped to 0th ground truth channel is first, etc., then evaluate the reordered predictions just like with supervised methods.

**Comment by Bralio123** [comment]
That makes much more sense now. 

Since each image produces HxW predictions, is it reasonable to find the one-to-one mapping for each image using all HxW predictions? e.g. for batch size 1 and H, W = 200 would produce 40,000 predictions.

**Comment by xu-ji** [comment]
Yes. For segmentation evaluation you can vectorise all the heatmap predictions into shape (B * H * W) x C for batch size B and C clusters. Getting the argmax over the second dimension should yield a length B * H * W vector suitable for `_hungarian_match`. The targets would also be B * H * W.

**Comment by Bralio123** [comment]
That's perfect. Thanks for the quick response.

I'll close the issue again.
